lmdeploy/serve/utils.py

```diff
@@ -0,0 +1,253 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import asyncio
+from typing import Dict, List, Tuple, Union
+
+import numpy as np
+import torch
+from torch.nn.utils.rnn import pad_sequence
+
+from lmdeploy.utils import get_logger
+
+logger = get_logger('lmdeploy')
+
+InputIdsType = List[int]
+InputEmbsType = Union[None, List[Union[torch.Tensor, np.ndarray]]]
+InputEmbRngsType = Union[None, List[Tuple[int, int]]]
+PromptType = Union[str, List[Dict]]
+
+
+def _get_event_loop():
+    """get event loop."""
+    try:
+        event_loop = asyncio.get_event_loop()
+    except Exception:
+        logger.warning('Can not found event loop in current thread.'
+                       ' Create a new event loop.')
+        event_loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(event_loop)
+    return event_loop
+
+
+class LogitsMixin:
+    """Helper class to calculate logits and ppl."""
+
+    def prepare_inputs(self, prompts: Union[PromptType, List[PromptType]]):
+        if hasattr(self, '_convert_prompts'):
```

Comment:
So, we always apply the chat template for the VLM models, but we don't do it on LLMs, right?